spring/src/main/java/io/seata/spring/annotation/datasource/SeataDataSourceBeanPostProcessor.java

```diff
@@ -0,0 +1,108 @@
+ *  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package io.seata.spring.annotation.datasource;
+
+import javax.sql.DataSource;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
+
+import io.seata.common.exception.ShouldNeverHappenException;
+import io.seata.config.ConfigurationFactory;
+import io.seata.rm.datasource.DataSourceProxy;
+import io.seata.spring.util.SpringProxyUtils;
+import net.sf.cglib.proxy.Enhancer;
+import net.sf.cglib.proxy.MethodInterceptor;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.BeanUtils;
+import org.springframework.beans.BeansException;
+import org.springframework.beans.factory.config.BeanPostProcessor;
+
+import static io.seata.core.constants.ConfigurationKeys.DATASOURCE_AUTOPROXY;
+import static io.seata.core.constants.ConfigurationKeys.DATASOURCE_USE_JDK_PROXY;
+
+/**
+ * @author xingfudeshi@gmail.com
+ * @date 2019/12/26
+ * The type seata data source bean post processor
+ */
+public class SeataDataSourceBeanPostProcessor implements BeanPostProcessor {
+    private static final Logger LOGGER = LoggerFactory.getLogger(SeataDataSourceBeanPostProcessor.class);
+
+    @Override
+    public Object postProcessAfterInitialization(Object bean, String beanName) throws BeansException {
+        if (bean instanceof DataSource && !(bean instanceof DataSourceProxy) && ConfigurationFactory.getInstance().getBoolean(DATASOURCE_AUTOPROXY, false)) {
+            if (LOGGER.isInfoEnabled()) {
+                LOGGER.info("Auto proxy of [{}]", beanName);
+            }
+            return proxyDataSource(bean);
+        }
+        return bean;
+    }
+
+    @Override
+    public Object postProcessBeforeInitialization(Object bean, String beanName) {
+        if (bean instanceof DataSourceProxy && ConfigurationFactory.getInstance().getBoolean(DATASOURCE_AUTOPROXY, false)) {
+            throw new ShouldNeverHappenException("Auto proxy of DataSource can't be enabled as you've created a DataSourceProxy bean." +
+                "Please consider removing DataSourceProxy bean or disabling auto proxy of DataSource.");
+        }
+        return bean;
+    }
+
+    /**
+     * proxy data source
+     *
+     * @param originBean
+     * @return proxied datasource
+     */
+    private Object proxyDataSource(Object originBean) {
+        DataSourceProxy dataSourceProxy = DataSourceProxyHolder.get().putDataSource((DataSource) originBean);
+        if (ConfigurationFactory.getInstance().getBoolean(DATASOURCE_USE_JDK_PROXY, false)) {
```

Comment:
In 1.0.0, jdk proxy is default, recommendations for consistency.

Reply:
Based on the known usage, cglib proxy can be adapted to all third parties except Java melody. In addition, the JDK proxy needs to use interface injection, and there may be some code intrusion for some users' existing business.